File: ts/i18n/src/registry.ts
/**
 * `createLocaleRegistry` — the locale set, the URL contract, negotiation and the
 * translator, bound to one list of locales.
 *
 * The core's free functions (`localePath`, `splitLocalePath`, …) are this
 * factory applied to the generated five-locale set, so a surface with its own
 * languages — a French storefront that is `fr` by default, say — gets the exact
 * same behaviour over its own list instead of a hand-rolled copy that drifts.
 *
 * Zero-dep and DOM-free, like the rest of the core.
 */
import { formatPattern, type MessageValues } from "./format";

/**
 * A loaded catalogue: flat `key → pattern`. Flat rather than nested because the
 * drift checker diffs, hashes, and reports per fully-qualified key, and a nested
 * shape would make every one of those operations a tree walk for no gain at the
 * call site (`t("hero.title")` reads the same either way).
 */
export type Messages = Readonly<Record<string, string>>;

/**
 * Renders one string: `en` is the source-language copy as authored at this call
 * site, `key` is what a translated catalogue files it under.
 *
 * ```ts
 * t("hero.title", "Invest in the China+1 narrative")
 * ```
 */
export type Translate = (key: string, en: string, values?: MessageValues) => string;

/**
 * The structural slice of `Element` `localeOfElement` reads. Spelled out rather
 * than imported from the DOM lib so the core keeps type-checking — and shipping
 * — without one.
 */
export interface LangScope {
  getAttribute(name: string): string | null;
  closest(selectors: string): LangScope | null;
}

/** A non-empty, ordered list of locale codes — `as const` keeps the literals. */
export type LocaleList = readonly [string, ...string[]];

/** What {@link createLocaleRegistry} is built from. */
export interface LocaleRegistryConfig<Locales extends LocaleList> {
  /**
   * Every locale, in the order a language switcher offers them. These are the
   * URL segments and catalogue directory names, so keep them to bare language
   * codes; a regional `hreflang` belongs in {@link hreflang}.
   */
  locales: Locales;
  /** Each locale's name in that locale — what a switcher must show. */
  labels: Readonly<Record<Locales[number], string>>;
  /** The fallback for any reader we cannot place, and the authored source. */
  default: Locales[number];
  /**
   * Whether the default locale's URLs carry a `/<locale>` prefix too.
   *
   * `false` (the default) keeps EV's contract: the default locale lives at bare
   * paths, so already-indexed URLs never move. `true` gives every locale a
   * prefix — the right shape for a site that launches in several languages at
   * once and has no legacy URLs to protect.
   */
  prefixDefaultLocale?: boolean;
  /**
   * The `hreflang` tag per locale, for a site that targets a region rather than
   * a language (`fr` → `fr-FR`, or `fr-BE` for a Belgian storefront). A locale
   * left out advertises its bare code.
   */
  hreflang?: Readonly<Partial<Record<Locales[number], string>>>;
}

/**
 * One locale set with everything that depends on it. Every member is a plain
 * closure, so destructuring (`const { localePath } = registry`) is safe.
 */
export interface LocaleRegistry<L extends string> {
  /** Every locale, in switcher order. */
  readonly locales: readonly L[];
  /** Each locale's endonym. */
  readonly labels: Readonly<Record<L, string>>;
  /** The fallback locale, and the language `t()` call sites are authored in. */
  readonly defaultLocale: L;
  /** Whether the default locale's URLs are prefixed as well. */
  readonly prefixDefaultLocale: boolean;
  /** Narrowing guard for untrusted input — a URL segment, a cookie, a query param. */
  readonly isLocale: (value: unknown) => value is L;
  /** The path `locale` serves `path` at. */
  readonly localePath: (locale: L, path: string) => string;
  /**
   * Split a request path into its locale and the locale-free path beneath it.
   * An absent or unrecognised prefix reads as the default locale; never throws.
   */
  readonly splitLocalePath: (pathname: string) => { locale: L; path: string };
  /** Every locale's root-relative URL for one page, keyed by locale code. */
  readonly localeAlternates: (path: string, locales?: readonly L[]) => Record<string, string>;
  /** The `hreflang` tag a locale is advertised under. */
  readonly hreflangOf: (locale: L) => string;
  /**
   * Absolute URLs for one page keyed by `hreflang` tag, plus `x-default` — the
   * `alternates.languages` value Next's `generateMetadata` wants.
   */
  readonly languageAlternates: (
    path: string,
    siteUrl: string,
    locales?: readonly L[],
  ) => Record<string, string>;
  /** Best locale for an `Accept-Language` header — for suggesting, not serving. */
  readonly negotiate: (header: string | null | undefined, locales?: readonly L[]) => L;
  /** A {@link Translate} bound to one catalogue and locale. */
  readonly translator: (
    messages: Messages,
    locale: L,
    onMissing?: (key: string, locale: L) => void,
  ) => Translate;
  /** Format one ICU-subset pattern in `locale`. */
  readonly formatMessage: (pattern: string, locale: L, values?: MessageValues) => string;
  /** The locale a DOM subtree is written in, per `lang` inheritance. */
  readonly localeOfElement: (node: LangScope) => L;
}

/**
 * Build a {@link LocaleRegistry} over one list of locales.
 *
 * The locale type is the literal union of `locales`, so a registry for
 * `["fr", "en"]` rejects `localePath("de", …)` at compile time.
 *
 * @throws If `default` is not one of `locales`, or a locale is listed twice —
 *   both would make the URL contract ambiguous, so they fail at startup rather
 *   than on the first request that hits them.
 *
 * @example
 * ```ts
 * const i18n = createLocaleRegistry({
 *   locales: ["fr", "en"],
 *   labels: { fr: "Français", en: "English" },
 *   default: "fr",
 *   prefixDefaultLocale: true,
 *   hreflang: { fr: "fr-FR" },
 * });
 * i18n.localePath("fr", "/contact");   // "/fr/contact"
 * ```
 */
export function createLocaleRegistry<const Locales extends LocaleList>(
  config: LocaleRegistryConfig<Locales>,
): LocaleRegistry<Locales[number]> {
  type L = Locales[number];
  const locales: readonly L[] = config.locales;
  const defaultLocale: L = config.default;
  const prefixDefaultLocale = config.prefixDefaultLocale ?? false;
  const hreflangTags = config.hreflang;

  if (!locales.includes(defaultLocale)) {
    throw new Error(`default locale "${defaultLocale}" is not in [${locales.join(", ")}]`);
  }
  if (new Set(locales).size !== locales.length) {
    throw new Error(`locales listed more than once: [${locales.join(", ")}]`);
  }

  const isLocale = (value: unknown): value is L =>
    typeof value === "string" && (locales as readonly string[]).includes(value);

  const isPrefixed = (locale: L) => prefixDefaultLocale || locale !== defaultLocale;

  const localePath = (locale: L, path: string): string => {
    const clean = path.startsWith("/") ? path : `/${path}`;
    if (!isPrefixed(locale)) return clean;
    // "/" would otherwise yield "/ru/", and a trailing slash is a distinct URL to
    // a crawler — one canonical shape per page, so strip it.
    return clean === "/" ? `/${locale}` : `/${locale}${clean}`;
  };

  const splitLocalePath = (pathname: string): { locale: L; path: string } => {
    const clean = pathname.startsWith("/") ? pathname : `/${pathname}`;
    const slash = clean.indexOf("/", 1);
    const head = slash === -1 ? clean.slice(1) : clean.slice(1, slash);
    // An explicit `/<default>` segment is not a prefix when the default is
    // unprefixed: it stays part of the path, exactly as a stray segment would.
    if (!isLocale(head) || !isPrefixed(head)) return { locale: defaultLocale, path: clean };
    const rest = slash === -1 ? "/" : clean.slice(slash);
    return { locale: head, path: rest === "" ? "/" : rest };
  };

  const localeAlternates = (path: string, only: readonly L[] = locales): Record<string, string> =>
    Object.fromEntries(only.map(l => [l, localePath(l, path)]));

  const hreflangOf = (locale: L): string => hreflangTags?.[locale] ?? locale;

  const languageAlternates = (
    path: string,
    siteUrl: string,
    only: readonly L[] = locales,
  ): Record<string, string> => {
    const origin = siteUrl.replace(/\/+$/, "");
    const abs = (l: L) => `${origin}${localePath(l, path)}`;
    return {
      ...Object.fromEntries(only.map(l => [hreflangOf(l), abs(l)])),
      // What a crawler serves a reader whose language matches none of ours —
      // the same answer the site itself gives.
      "x-default": abs(defaultLocale),
    };
  };

  const negotiate = (header: string | null | undefined, only: readonly L[] = locales): L => {
    if (!header) return defaultLocale;

    const ranked = header
      .split(",")
      .map(part => {
        const [tag, ...params] = part.trim().split(";");
        const q = params
          .map(p => p.trim())
          .find(p => p.startsWith("q="))
          ?.slice(2);
        const quality = q === undefined ? 1 : Number.parseFloat(q);
        return {
          tag: (tag ?? "").trim().toLowerCase(),
          // A malformed q= sorts last rather than poisoning the comparison with NaN.
          quality: Number.isFinite(quality) ? quality : 0,
        };
      })
      // q=0 is an explicit refusal of that language, not a weak preference.
      .filter(entry => entry.tag !== "" && entry.quality > 0)
      .sort((a, b) => b.quality - a.quality);

    for (const { tag } of ranked) {
      // "ru-RU" and "ru" both match the "ru" catalogue; "*" means "anything", for
      // which the default is as good an answer as any.
      const base = tag.split("-")[0] ?? "";
      const hit = only.find(l => {
        const code = l.toLowerCase();
        return code === tag || code === base;
      });
      if (hit) return hit;
      if (tag === "*") return defaultLocale;
    }
    return defaultLocale;
  };

  const formatMessage = (pattern: string, locale: L, values?: MessageValues): string =>
    formatPattern(pattern, locale, values);

  const translator = (
    messages: Messages,
    locale: L,
    onMissing?: (key: string, locale: L) => void,
  ): Translate => {
    // The default locale is the authored source: its copy is the call site's.
    const source = locale === defaultLocale ? null : messages;
    return (key, en, values) => {
      if (source === null) return formatPattern(en, locale, values);
      const pattern = source[key];
      if (pattern === undefined) onMissing?.(key, locale);
      return formatPattern(pattern ?? en, locale, values);
    };
  };

  const localeOfElement = (node: LangScope): L =>
    negotiate(node.closest("[lang]")?.getAttribute("lang"));

  return {
    locales,
    labels: config.labels,
    defaultLocale,
    prefixDefaultLocale,
    isLocale,
    localePath,
    splitLocalePath,
    localeAlternates,
    hreflangOf,
    languageAlternates,
    negotiate,
    translator,
    formatMessage,
    localeOfElement,
  };
}

File: rust/src/i18n/registry.rs
//! A locale registry over a caller's own list — the Rust half of TS
//! `createLocaleRegistry`.
//!
//! The free functions in [`super`] are fixed to EV's five locales and to
//! "the default is unprefixed". A surface with its own languages — a French
//! storefront that is `fr` by default and prefixes every locale — gets the same
//! URL contract and the same negotiation here instead of a hand-rolled copy
//! that drifts. Locales are plain codes (`"fr"`), validated once at
//! construction; every accessor then answers with the registry's own `&str`.

use super::ranked_tags;

/// What a [`LocaleRegistry`] is built from.
///
/// ```
/// use ev_lib::i18n::{LocaleRegistry, LocaleRegistryConfig};
/// let i18n = LocaleRegistry::try_new(LocaleRegistryConfig {
///     locales: vec![("fr".into(), "Français".into()), ("en".into(), "English".into())],
///     default: "fr".into(),
///     prefix_default_locale: true,
///     hreflang: vec![("fr".into(), "fr-FR".into())],
/// })
/// .unwrap();
/// assert_eq!(i18n.locale_path("fr", "/contact"), "/fr/contact");
/// assert_eq!(i18n.hreflang_of("fr"), "fr-FR");
/// assert_eq!(i18n.negotiate(Some("en-GB,fr;q=0.5")), "en");
/// ```
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct LocaleRegistryConfig {
	/// `(code, label)` in the order a language switcher offers them. The code is
	/// the URL segment, so keep it a bare language; the label is the locale's
	/// name in that locale.
	pub locales: Vec<(String, String)>,
	/// The fallback for any reader we cannot place.
	pub default: String,
	/// Whether the default locale's URLs carry a `/<locale>` prefix too.
	pub prefix_default_locale: bool,
	/// `(code, tag)` for a locale advertised under a regional `hreflang`
	/// (`fr` → `fr-FR`). A locale left out advertises its bare code.
	pub hreflang: Vec<(String, String)>,
}

/// Why a [`LocaleRegistryConfig`] was refused. Each would make the URL contract
/// ambiguous, so it fails at startup rather than on the first request.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum RegistryError {
	NoLocales,
	DefaultNotListed(String),
	ListedTwice(String),
	/// An `hreflang` entry for a code that is not one of the locales.
	UnknownHreflangLocale(String),
	/// `x-default` names the fallback; no locale may be advertised under it.
	ReservedHreflang(String),
	/// Two locales advertised under one tag collapse into one alternate.
	HreflangCollision(String),
}

impl std::fmt::Display for RegistryError {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Self::NoLocales => f.write_str("a locale registry needs at least one locale"),
			Self::DefaultNotListed(code) => write!(f, "default locale {code:?} is not one of the locales"),
			Self::ListedTwice(code) => write!(f, "locale {code:?} is listed more than once"),
			Self::UnknownHreflangLocale(code) => write!(f, "hreflang is given for {code:?}, which is not one of the locales"),
			Self::ReservedHreflang(code) => write!(f, "locale {code:?} cannot be advertised as \"x-default\""),
			Self::HreflangCollision(tag) => write!(f, "hreflang tag {tag:?} is used by more than one locale"),
		}
	}
}

impl std::error::Error for RegistryError {}

/// One locale set with everything that depends on it: the URL contract,
/// `hreflang` clusters and `Accept-Language` negotiation.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct LocaleRegistry {
	entries: Vec<Entry>,
	default: usize,
	prefix_default_locale: bool,
}
impl LocaleRegistry {
	/// Validate `config` and build the registry.
	pub fn try_new(config: LocaleRegistryConfig) -> Result<Self, RegistryError> {
		if config.locales.is_empty() {
			return Err(RegistryError::NoLocales);
		}
		let mut entries: Vec<Entry> = Vec::with_capacity(config.locales.len());
		for (code, label) in config.locales {
			if entries.iter().any(|e| e.code == code) {
				return Err(RegistryError::ListedTwice(code));
			}
			entries.push(Entry {
				hreflang: code.clone(),
				code,
				label,
			});
		}
		for (code, tag) in config.hreflang {
			let entry = entries.iter_mut().find(|e| e.code == code).ok_or(RegistryError::UnknownHreflangLocale(code))?;
			entry.hreflang = tag;
		}
		// Checked on the effective tags, bare-code fallbacks included.
		for (i, entry) in entries.iter().enumerate() {
			if entry.hreflang.eq_ignore_ascii_case("x-default") {
				return Err(RegistryError::ReservedHreflang(entry.code.clone()));
			}
			if entries[..i].iter().any(|e| e.hreflang.eq_ignore_ascii_case(&entry.hreflang)) {
				return Err(RegistryError::HreflangCollision(entry.hreflang.clone()));
			}
		}
		let default = entries.iter().position(|e| e.code == config.default).ok_or(RegistryError::DefaultNotListed(config.default))?;
		Ok(Self {
			entries,
			default,
			prefix_default_locale: config.prefix_default_locale,
		})
	}

	/// Every locale, in switcher order.
	pub fn locales(&self) -> impl ExactSizeIterator<Item = &str> + '_ {
		self.entries.iter().map(|e| e.code.as_str())
	}

	pub fn default_locale(&self) -> &str {
		&self.entries[self.default].code
	}

	pub fn prefix_default_locale(&self) -> bool {
		self.prefix_default_locale
	}

	/// Narrow untrusted input — a URL segment, a cookie, a query param — to one
	/// of the registry's locales.
	pub fn locale(&self, value: &str) -> Option<&str> {
		self.entry(value).map(|e| e.code.as_str())
	}

	pub fn is_locale(&self, value: &str) -> bool {
		self.entry(value).is_some()
	}

	/// The locale's name in that locale — what a switcher must show.
	pub fn label(&self, locale: &str) -> Option<&str> {
		self.entry(locale).map(|e| e.label.as_str())
	}

	/// The `hreflang` tag a locale is advertised under; an unknown code is
	/// advertised as itself.
	pub fn hreflang_of<'a>(&'a self, locale: &'a str) -> &'a str {
		self.entry(locale).map_or(locale, |e| e.hreflang.as_str())
	}

	/// The path `locale` serves `path` at.
	pub fn locale_path(&self, locale: &str, path: &str) -> String {
		let clean = if path.starts_with('/') { path.to_owned() } else { format!("/{path}") };
		if !self.is_prefixed(locale) {
			return clean;
		}
		// "/" would otherwise yield "/fr/", and a trailing slash is a distinct URL
		// to a crawler.
		if clean == "/" { format!("/{locale}") } else { format!("/{locale}{clean}") }
	}

	/// Split a request path into its locale and the locale-free path beneath
	/// it. An absent or unrecognised prefix reads as the default locale.
	pub fn split_locale_path(&self, pathname: &str) -> (&str, String) {
		let clean = if pathname.starts_with('/') { pathname.to_owned() } else { format!("/{pathname}") };
		let after = &clean[1..];
		let (head, rest) = after.find('/').map_or((after, ""), |i| (&after[..i], &after[i..]));
		match self.locale(head) {
			// An explicit `/<default>` segment is not a prefix when the default is
			// unprefixed: it stays part of the path, like any stray segment.
			Some(locale) if self.is_prefixed(locale) => (locale, if rest.is_empty() { "/".to_owned() } else { rest.to_owned() }),
			_ => (self.default_locale(), clean),
		}
	}

	/// Every locale's root-relative URL for one page, in switcher order.
	pub fn locale_alternates(&self, path: &str) -> Vec<(&str, String)> {
		self.locales().map(|l| (l, self.locale_path(l, path))).collect()
	}

	/// Absolute URLs for one page keyed by `hreflang` tag, in switcher order,
	/// then `x-default` on the default locale's URL.
	pub fn language_alternates(&self, path: &str, site_url: &str) -> Vec<(String, String)> {
		let origin = site_url.trim_end_matches('/');
		let mut out: Vec<(String, String)> = self
			.entries
			.iter()
			.map(|e| (e.hreflang.clone(), format!("{origin}{}", self.locale_path(&e.code, path))))
			.collect();
		out.push(("x-default".to_owned(), format!("{origin}{}", self.locale_path(self.default_locale(), path))));
		out
	}

	/// Best locale for an `Accept-Language` header — for suggesting, not serving.
	pub fn negotiate(&self, header: Option<&str>) -> &str {
		let Some(header) = header else {
			return self.default_locale();
		};
		for tag in ranked_tags(header) {
			let base = tag.split('-').next().unwrap_or("");
			if let Some(hit) = self.entries.iter().find(|e| {
				let code = e.code.to_ascii_lowercase();
				code == tag || code == base
			}) {
				return &hit.code;
			}
			if tag == "*" {
				return self.default_locale();
			}
		}
		self.default_locale()
	}

	fn entry(&self, code: &str) -> Option<&Entry> {
		self.entries.iter().find(|e| e.code == code)
	}

	fn is_prefixed(&self, locale: &str) -> bool {
		self.prefix_default_locale || locale != self.default_locale()
	}
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct Entry {
	code: String,
	label: String,
	hreflang: String,
}
